src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,290 @@
+                    {
+                        return projectInfo.SymbolTreeInfo;
+                    }
                 }
+
+                return null;
+            }
+        }
+
+        private class IncrementalAnalyzer : IncrementalAnalyzerBase
+        {
+            private readonly ConcurrentDictionary<ProjectId, ProjectInfo> _projectToInfo;
+
+            // Note: the Incremental-Analyzer infrastructure guarantees that it will call all the methods
+            // on this type in a serial fashion.  As such, we don't need explicit locking, or threadsafe
+            // collections (if they're only used by this type).  So, for example, the map we populate
+            // needs to be a ConcurrentDictionary as it will be read and written from multiple types.
+            // However, the HashSet<ProjectId> is ok as it will only be used by this type and there is
+            // no concurrency in this type on its own.
+            private readonly ConcurrentDictionary<string, MetadataInfo> _metadataPathToInfo;
+
+            private readonly Dictionary<ProjectId, Project> idTolastSeenProject = new Dictionary<ProjectId, Project>();
+
+            public IncrementalAnalyzer(
+                ConcurrentDictionary<ProjectId, ProjectInfo> projectToInfo,
+                ConcurrentDictionary<string, MetadataInfo> metadataPathToInfo)
+            {
+                _projectToInfo = projectToInfo;
+                _metadataPathToInfo = metadataPathToInfo;
             }
 
-            public override Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken)
+            public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
             {
-                // check whether we are good to report total symbol numbers
-                if (_symbolCountByProjectMap == null || _symbolCountByProjectMap.Count < solution.ProjectIds.Count || string.IsNullOrEmpty(solution.FilePath))
+                if (!document.SupportsSyntaxTree)
                 {
+                    // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                if (_solutionId != null && _solutionId != solution.Id)
+                if (bodyOpt != null)
                 {
-                    ReportCount();
+                    // This was a method level edit.  This can't change the symbol tree info
+                    // for this project.  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
+            }
+
+            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
```

Comment:
Oh, we're keeping this one around as well? Is this made redundant by the document-triggered version above?